### ISSUE TITLE
Unhelpful type error using Converter where StreamTransformer expected

### ISSUE THREAD

**Issue by skybrian** [dart-lang/sdk#25698]
This error message repeats itself. I'm also not sure what I could do about it.

```
149: [ERROR] Type check failed: new Utf8Decoder() (Utf8Decoder) is not of type StreamTransformer<List<int>, dynamic> because new Utf8Decoder() cannot be typed as StreamTransformer<List<int>, dynamic> [STRONG_MODE_STATIC_TYPE_ERROR]
```

The code seems pretty straightforward:

```
    _process.stdout
        .transform(new Utf8Decoder())
        .transform(new LineSplitter())
        .listen((String line) { .. }
```

**Comment by jmesserly** [comment]
Could swear we had a bug for this, but I can't find it. In any case, yeah we need to fix that message. @munificent might be able to take a look?

**Comment by munificent** [comment]
The error message can definitely be improved, but it also appears fish. A simpler standalone repro is:

```dart
import 'dart:async';
import 'dart:convert';

main(Stream<List<int>> stream) {
  stream.transform(new Utf8Decoder());
}
```

Let's walk through the types:

* [`Stream<T>.transform()`](https://api.dartlang.org/1.14.2/dart-async/Stream/transform.html) expects a `StreamTransformer<T, dynamic>`.

* [Utf8Decoder](https://api.dartlang.org/1.14.2/dart-convert/Utf8Decoder-class.html) extends `Converter<List<int>, String>`.

* [Converter<S, T>](https://api.dartlang.org/1.14.2/dart-convert/Converter-class.html) implements `StreamTransformer` *without any type parameters*. Shouldn't it implement `StreamTransformer<T, S>`?

If I cobble together a synthetic example that fixes that:

```dart
class MyTransformer<S, T> {}
class MyConverter<S, T> implements MyTransformer<S, T> {} // <--
class MyDecoder extends MyConverter<List<int>, String> {}
class MyStream<T> {
  void transform(MyTransformer<T, dynamic> transformer) {}
}

main(MyStream<List<int>> stream) {
  stream.transform(new MyDecoder());
}
```

Then indeed the error goes away. @lrhn or @floitschG, is the core library wrong? Can I fix it?

**Comment by munificent** [comment]
cc @vsmenon @leafpetersen

**Comment by floitschG** [comment]
This was a deliberate choice. The stream-transformer doesn't always use the same source and output type.
(See https://codereview.chromium.org/1698563002/ )

There is no way to fix this, without adding more generic types to codecs, which we decided not to do when we designed the API.

**Comment by munificent** [comment]
> The stream-transformer doesn't always use the same source and output type.

Yes, I see that, but Converter is one-way anyway, so that doesn't seem like a problem to me. I would expect `bind()` to be `Stream<T> bind(Stream<S> source)` in Converter, but it isn't. What am I missing?

> There is no way to fix this, without adding more generic types to codecs

Is this because of `fuse()`?

> which we decided not to do when we designed the API.

Time to revisit that decision?

**Comment by floitschG** [comment]
Let me rephrase: The stream-transformer doesn't always use the same source and output type as the codec.

You just LGTMed a CL where that's not the case.

**Comment by munificent** [comment]
> The stream-transformer doesn't always use the same source and output type as the codec.

I'm not sure what you mean by "the stream-transformer" and "the codec".  Are these the same objects? There are no Codecs mentioned by this issue, just Converter.

> You just LGTMed a CL where that's not the case.

Yes, though I admit I didn't fully understand the change. :-/

Let me try asking this a different way. With generic methods, can we make this API type safe? Should we?

I don't know streams very well, but my expectation is that the classes in question would look something like:

```dart
class Stream<S> {
  Stream<T> transform<T>(StreamTransformer<S, T> transformer) {
    return streamTransformer.bind(this);
  }
}

abstract class StreamTransformer<S, T> {
  Stream<T> bind(Stream<S> source);
}

abstract class Converter<S, T> implements StreamTransformer<S, T> {
  Stream<T> bind(Stream<S> source) => source.map(convert);

  T convert(S input);

  Converter<S, U> fuse<U>(Converter<T, U> other) { ... }
}

class LineSplitter implements Converter<String, List<String>> {
  List<String> convert(String input) => input.split("\n");
}
```

Does this look right?

**Comment by floitschG** [comment]
It is simply not true that `abstract class Converter<S, T> implements StreamTransformer<S, T>`

For LineSplitter it would be:
`class LineSplitter implements Converter<String, List<String>>, StreamTransformer<String, String>`

Therefore `Converter` can not set the generic types of `StreamTransformer`. A subclass of `Converter` inherits `StreamTransformer` from `Converter` and can't update the type anymore. It can "only" update the type of the `bind` method (which is, what we do).

**Comment by vsmenon** [comment]
It looks like the simplest fix here is to have:

```dart
class Converter<S, T> implements StreamTransformer<S, dynamic>
```

`Converter<S,T>.bind` takes the same input type - and we can validate that - but does not promise to return the same return type as `StreamTransformer<S,T>.bind`.

Is that true in general?

**Comment by floitschG** [comment]
No. It isn't.

It's not the best example, but a `LineConcatenater` would take `List<String>` for `convert`, but take simple Strings otherwise.

**Comment by munificent** [comment]
> It's not the best example, but a LineConcatenater would take List<String> for convert, but take simple Strings otherwise.

Given these cases, why does Converter claim that it implements StreamTransformer at all? Or, if it's going to, why not provide distinct type parameters for them?

What's the type safe way for us to untangle this?

**Comment by vsmenon** [comment]
I agree with Bob.  If we want to support that sort of example, `Converter` should not implement `StreamTransformer`.

Much cleaner to say:
```dart
LineConcatenater implements  Converter<List<String>, String>, StreamTransformer<String, String>
```
if that is the intent.  Why not make this change?

**Comment by floitschG** [comment]
Because a converter has a `chunkedConversion` (which is basically the synchronous version of the StreamTransformer). The `chunkedConversion` is not part of the `StreamTransformer` interface, but the `Converter` interface.

**Comment by vsmenon** [comment]
If it's not in `StreamTransformer` how does it affect this?

**Comment by floitschG** [comment]
Because the Baseclass of `Converter` provides a default-implementation of `StreamTransformer` based on `chunkedConversion`. Since there is a default-implementation, it only makes sense that it also advertises that capability.

**Comment by munificent** [comment]
> Because the baseclass of Converter provides a default-implementation of StreamTransformer based on chunkedConversion.

That makes sense.

> Since there is a default-implementation, it only makes sense that it also advertises that capability.

In that case, it would be great if it advertised the correct types that capability provides and requires. Should we add more type parameters to Converter so it can distinguish between its `bind()` interface and its `convert()` one? Treating them as having the same type is obviously wrong, but treating them as completely untyped isn't very helpful either.

**Comment by vsmenon** [comment]
I don't see a `chunkedConversion` method on `Converter`:

https://github.com/dart-lang/sdk/blob/master/sdk/lib/convert/converter.dart

There is a similarly named method, but it just throws.

**Comment by floitschG** [comment]
Yes. It's `startChunkedConversion`.

**Comment by vsmenon** [comment]
It just throws ... why do we need to advertise that capability?  I'm not following you.

**Comment by floitschG** [comment]
The `Converter` class is an abstract class. The `startChunkedConversion` function should be implemented by the converter if it supports chunked conversions (and thus stream transformations). If it doesn't it will throw at runtime. (This is similar to some lists not supporting changing their size).

If implemented, `startChunkedConversion` will be used by the base-implementation of the `StreamTransformer` functionality. This means that the synchronous `startChunkedConversion` is used by the asynchronous `StreamTransformer`, and since there is a default implementation of the `StreamTransformer` functions, `Converter` implements `StreamTransformer`.

The type-wise sound thing would be to pass 4 generic types to a `Converter`: 2 for the non-chunked conversion, and 2 for the chunked conversion.
Initially, we implemented converters this way, but soon discovered that the additional 2 arguments are too cumbersome. Not only do users generally think of Converters going from `S` to `T` (and not going from `S` to `T`, but also from `S2` to `T2`), the additional arguments also don't apply to converters that don't implement the chunked conversion. Given our experience with the 4 generic arguments, we made the deliberate change to go to 2 generic types and setting the (static) type of the StreamTransformer to dynamic->dynamic.
Although we did remove the 2 generic types for the chunked conversion, we encourage implementers to override the `bind` method to give it a more precise type. This means that, dynamically, most bad uses of transformers are caught.
I would have prefered if subclasses can tighten their generic types (so that the analyzer could more easily see what's going on), but dart2js doesn't support multiple implements of the same type. Also, according to the semantics the interface types would be combined and not tightened.

``` dart
class A<T> {}
class B extends A {}
class C extends B implements A<int> {}

main() {
  print(new C() is A<int>);  // => true.
  print(new C() is A<String>);  // => true.
}
```

In any case, even if I wanted to change the code it's not something we could do. Adding 2 more type-arguments to the `Converter` class is a breaking change, as would be removing the `StreamTransformer` interface from `Converter`.

**Comment by skybrian** [comment]
It sounds like the strong mode type system needs to special-case stream converters (to preserve backward compatibility).

**Comment by vsmenon** [comment]
Yes, we'll need to either:

1. Expand the type system to capture this example.
2. Find a minimal & acceptable breaking change to type this.

We'll explore option 1 first and update here.  Longer term, we may want to find a better typing regardless.

**Comment by munificent** [comment]
OK, I've made the error message better here, which is all I plan to do right now, so I'm going to unassign myself and update the title.

**Comment by skybrian** [comment]
Thanks. In the meantime, is there a workaround? Some other way of writing the code?

**Comment by jmesserly** [comment]
I'm not sure, but this feels to me like high priority. I believe @vsmenon was looking into some type system workarounds.

**Comment by vsmenon** [comment]
@skybrian : This should suppress the static warning for now.

```dart
    _process.stdout
        .transform(new Utf8Decoder() as StreamTransformer<List<int>, dynamic>)
        .transform(new LineSplitter() as StreamTransformer<String, dynamic>)
        .listen((String line) { .. }
```

**Comment by nex3** [comment]
This is blocking making the `crypto` package strong-mode clean. We have several APIs like the following:

```dart
abstract class Hash extends Converter<List<int>, Digest> {
  ByteConversionSink startChunkedConversion(Sink<Digest> sink);
}
```

This fails in strong mode because `(Sink<Digest>) → ByteConversionSink` is not a subtype of `(Sink<dynamic>) → ChunkedConversionSink<dynamic>`.

**Comment by nex3** [comment]
As for the larger issue: as far as I can tell, `LineSplitter` is the *only* instance where we have a `Converter<S, T>` that isn't also a `StreamTransformer<S, T>`. And the only reason there's a difference there is so that `LineSplitter.convert()` can return a `List<String>`—and as far as I know, that method is entirely redundant with the `LineSplitter.split()` static method anyway. So it looks like we don't actually get any concrete value from having `Converter<S, T>` not implement `StreamTransformer<S, T>`.

I propose that we make `LineSplitter.convert()` return a single string—either the first line, or the lines split and rejoined in a more normalized way, whatever—and make `Converter<S, T>` implement `StreamTransformer<S, T>` which makes the rest of the world much simpler.

**Comment by kevmoo** [comment]
Removing `analyzer` labels on this issue. I think the analyzer is doing the right thing here – we need to revisit this API design

**Comment by nex3** [comment]
My understanding was that the analyzer-strong-mode label was being used in part to track issues that are blocking strong mode adoption, which this is.

**Comment by jmesserly** [comment]
yeah. I wonder if we should rename analyzer-strong-mode to just strong-mode?

**Comment by dgrove** [comment]
@floitschG any updates?

**Comment by dgrove** [comment]
@floitschG update please?

**Comment by floitschG** [comment]
With 50bdab3841dafd9b056b1dd013c103ce6a566f60 the converters can have full type information.
The patch is a transitional patch to give users time to adjust their code (since it's a breaking change).

The existing converters (of the core libraries) are adapted.

**Comment by nex3** [comment]
@floitschG Why doesn't `ChunkedConverter` implement `StreamTransformer<S2, T2>`?

**Comment by floitschG** [comment]
@nex3: that is where we want to get to. However, we can't just break everyone in a dot-release.

The conversion library is now set up in a way that users can migrate their code. In a future release we will then do the breaking change and make `Converter` not extend `StreamTransformer` (and instead make `ChunkedConverter` extend it).

I will send out a migration mail when 1.16 is released.

**Comment by leafpetersen** [comment]
@floitschG I think this is fixed in the latest dev release, is that correct?

**Comment by jmesserly** [comment]
Just verified this is fixed:
```dart
import 'dart:io';
import 'dart:convert';

main() async {
  var p = await Process.start('ls', ['-l']);
  p.stdout
        .transform(new Utf8Decoder())
        .transform(new LineSplitter())
        .listen((String line) { print(line); });
  await p.exitCode;
}
```
```
$ dartanalyzer --strong test.dart 
Analyzing test.dart...
No issues found!
```
